rust/build.rs

```diff
@@ -0,0 +1,49 @@
+//! This build script is calculates the hash of all files in the `src/`
+//! directory and adds it as an environment variable during build time.
+//!
+//! This is used so that the python code can detect when the built native module
+//! does not match the source in-tree, helping to detect the case where the
+//! source has been updated but the library hasn't been rebuilt.
+
+use std::path::PathBuf;
+
+use blake2::{Blake2b512, Digest};
+
+fn main() -> Result<(), std::io::Error> {
+    let mut dirs = vec![PathBuf::from("src")];
+
+    let mut paths = Vec::new();
+    while let Some(path) = dirs.pop() {
+        let mut entries = std::fs::read_dir(path)?
+            .map(|res| res.map(|e| e.path()))
+            .collect::<Result<Vec<_>, std::io::Error>>()?;
+
+        entries.sort();
+
+        let mut new_dirs = Vec::new();
+
+        for entry in entries {
+            if entry.is_dir() {
+                new_dirs.push(entry);
+            } else {
+                paths.push(entry.to_str().expect("valid rust paths").to_string());
+            }
+        }
+
+        dirs.append(&mut new_dirs);
```

Comment:
As someone who has written fewer than 10 lines of Rust, why can't we append to `dirs` directly inside the loop?

Reply:
Woops, this was a holdover from a previous iteration of the code. Will move it inside!